### PR TITLE
#174566160 Build user account endpoint

### PR DESCRIPTION
**Pull request by Extremophile-ai**
## What does this PR do ##
This PR creates a user account

## Description of tasks to be completed? ##
- Install relevant dependencies
- Initialize them
- Write program to create a user
- Test program to ensure user is created successfully

## How should this be manually tested/checked? ##
- Clone the PR
- Cd into the branch `ft-user-sign-up-174566160`
Create a .env file and define your database as follows:
`DEV_DATABASE_URL = <development database url>`
`TEST_DATABASE_URL = <test database url>`
`PROD_DATABASE_URL = <production database url>`
- Also, in the .env file, set `JWT_KEY = secret`
- Install the dependencies using npm install
- To test all endpoints, `npm run test` comes handy.
- Start the server by running `npm run start-dev`
- Test the post route using using Postman with `http://localhost:3000/api/v1/users/signup` as url.
- A user must sign up with a username, an email and a password

## What are the relevant pivotal tracker stories? ##
[#174566160](https://www.pivotaltracker.com/story/show/174566160)

## Architectural flowchart ##
[Create User google docs](https://docs.google.com/document/d/1OMtfRXlA0SqvPj9OMpScc94U0c9sg07qmXCEVtDxZ54/edit?usp=sharing)

### CODE REVIEW

**Review by AbonyiXavier — COMMENTED** [review]
1. Please you need to check for a duplicate of email been signed up with using sequelize `findOne` method, just to avoid same email been used.

2. Please remove your test from index-test file, create a `user.js` file and put your signup test inside

**Comment by Extremophile-ai** [comment]
> 1. Please you need to check for a duplicate of email been signed up with using sequelize `findOne` method, just to avoid same email been used.
> 2. Please remove your test from index-test file, create a `user.js` file and put your signup test inside

Alright

**Review by codeliezel — COMMENTED** [review]
- the Jwt.js file should be in utilities.
- the user service file should be in services folder, create a services folder in the src folder and have something like this;

**src/services/userService.js**

- why don't you have an error 409 email conflict test?
- create a controllers file in tests and add controllers test there, also move all models tests in a folder called models..

Good job for now...

File: src/UserService/Jwt.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import jwt from "jsonwebtoken";
```
[the remark this answers is not in the document]
Comment (Extremophile-ai) [reply src/UserService/Jwt.js:1]: Ok...

File: src/UserService/User.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import database from "../models";
```
[the remark this answers is not in the document]
Comment (Extremophile-ai) [reply src/UserService/User.js:1]: Ok.
Done

**Review by Augusta-E — DISMISSED** [review]

**Review by donaldcrane — APPROVED** [review]

**Review by Fiyin-Anne — APPROVED** [review]

**Review by Fiyin-Anne — APPROVED** [review]

**Review by Augusta-E — DISMISSED** [review]

**Review by bellogo — APPROVED** [review]

**Review by Endowmissy — COMMENTED** [review]
@Extremophile-ai in the seed file, encrypt password.
work on username and password validation because as a user I can sign up if I put spaces in my username and password, so that needs to be fixed, your validation should also fix extra spaces
return all messages at once and it should be in JSON format

**Comment by Extremophile-ai** [comment]
> @Extremophile-ai in the seed file, encrypt password.
> work on username and password validation because as a user I can sign up if I put spaces in my username and password, so that needs to be fixed, your validation should also fix extra spaces
> return all messages at once and it should be in JSON format

Ok.

**Review by Fiyin-Anne — APPROVED** [review]

**Review by donaldcrane — APPROVED** [review]

**Review by Bernard-Namangala — DISMISSED** [review]

**Review by codeliezel — CHANGES_REQUESTED** [review]
When I try to sign up different users, it asks for different secret keys, I don't like that, pls change it. Also, create a sample test format on your PR on how to test the API, where to put our secret ket in env, all that info..... add it.

**Comment by codeliezel** [comment]
Ok i get why must JWT_KEY = secret instead of secret = <a value>

**Comment by Extremophile-ai** [comment]
> Ok i get why must JWT_KEY = secret instead of secret =

Yes, the JWT_KEY = secret should be in the .env file.
and it is there on the PR description

**Review by Fiyin-Anne — APPROVED** [review]

**Review by donaldcrane — APPROVED** [review]

**Review by codeliezel — APPROVED** [review]

**Pull request merged**